### ISSUE TITLE
Missing error on recursive reference in local function return type inference

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#45942]
Null safety [changes](https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md#local-return-type-inference) how return types of local functions are inferred:

> With null safety, local function body inference is changed so that the local function name is not considered available for inference while performing inference on the body. As a result, any recursive calls to the function for which the result type is required for inference to complete will no longer be treated as having return type `dynamic`, but will instead result in an inference failure.

According to that, this program should have a compile-time error:

```dart
void main() {
  g(bool c) {
    if (c) {
      return g(c);
    } else {
      return 3;
    }
  }
}
```

Unfortunately, neither CFE nor analyzer report an error. That's a bug according to the spec. The bug slipped through because [the test for this wasn't run by the test runner](https://dart-review.googlesource.com/c/sdk/+/198141/3/tests/language/nnbd/static_errors/local_function_inference_test.dart#28). :(

Fixing it now is probably a breaking change. Filing this issue to discuss whether we want to ship that breaking change or not and, if so, to track the implementation work.

cc @leafpetersen @stereotype441 @lrhn @natebosch @eernstg @jakemac53

**Comment by natebosch** [comment]
Why was that part added to the spec, what would be the consequences of not fixing this?

Would we be disallowing useful behavior by making this an error?

**Comment by munificent** [comment]
> Why was that part added to the spec

I assume the goal was to eliminate a corner where an unexpected `dynamic` could sneak in.

> what would be the consequences of not fixing this?

Users might be getting `dynamic` in places where they don't expect. The impact is probably minimal since apparently know one even noticed until I discovered the test wasn't being run.

> Would we be disallowing useful behavior by making this an error?

Not unless you think implicit `dynamic` is useful.

**Comment by eernstg** [comment]
Here's a sketch of an approach that could make the breakage smaller and preserve static safety:

For a local function `f` with no declared return type, the name `f` is made available during a "trial" inference on the body with assumed return type `Never`, using the context type `_` for inference of returned expressions. If the inference succeeds yielding a return type `R` then a second run of inference is performed on the body assuming `R` as the return type of `f` and as the context type for returned expressions (the second run starts from scratch, so we may need to erase the results from the first run, in order to treat `f` as if its return type were declared as `R`).

There may be some pathological cases where this approach yields unwanted results, but there would presumably be a large number of cases where it enables return type inference for recursive functions in a useful manner. In the cases where it fails we just get an error, as we would with the currently specified approach.

**Comment by lrhn** [comment]
@eernstg So the idea is to run once to get the return type, then run again (only if we actually *used* the `Never` return type of the function for anything which could fail, so not just assigning to `void`) to get a correct static typing of the code that depends on the return type?
There probably are pathological examples involving LUBs of the return type. Say:
```dart
f(x) => x < 0 ? [0] : [f(x - 1)];
```
where the first go would find `LUB(List<int>, List<Never>)` == `List<int>`, but the actual sound result should be `List<Object>`.
(But as you say, if the second run doesn't match, we can still make it an error).

**Comment by eernstg** [comment]
Exactly, that's a good case in the sense that it will emit an error when we try out the candidate return type `List<int>`, and that sends the signal to the developer that "this is too hard, please help me!" as it should. ;-)

**Comment by srawlins** [comment]
Quick update that analyzer and CFE still each accept the program.
